File: src/core/messages/Msg.ts
import type { Remoji } from "./Remoji";

/**
 * Base template for a chat message.
 */
export interface Msg {
  /** The sender of this message. */
  sender: String,
  /** The content of this message. */
  content: String,
  /** The time at which this message was send. */
  time: String,
  
  /** Reaction emojis. ``(undefined if there are no emoji reactions)`` */
  remojis?: Array<Remoji> | undefined,

  /** The target message of this message. ``(undefined if not a reaction)`` */
  target?: Msg | undefined,

  /** Whether this message is a ping for the local user. ``(undefined if false)`` */
  ping?: Boolean | undefined,
}

/**
 * Extension for thread messages.
 */
export interface ThreadMsg extends Msg {
  /** The messages send in this thread. */
  messages: Array<Msg>,

  /* NOTE: The title of the thread is stored within the ``Msg -> content`` */
}

File: src-tauri/src/history/msg.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum Msg {
  /// Normal text message
  Text(Text),
  /// Inline reaction message
  InlineReaction(InlineReaction),
  // Image, Embeds, Invites, etc...
}

#[derive(Clone, Serialize, Deserialize, Debug)]
/// # Text
/// The default chat message.
/// These contain text and a sender.
///
/// # Format
/// ```
/// @sender : text message content
/// ```
pub struct Text {
  /** ID of the user who send this message */
  sender: u8,
  /** Content of the message */
  content: String,
  /** Time at which the message was send */
  // time: u64, // This should be a DateTime<UTC>

  /** Reaction emojis */
  remojis: Vec<Remoji>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
/// # Inline Reactions
/// A reaction to another message in the same chat.
/// These reactions are inserted at the bottom of the chat.
/// Just like normal text messages except the refer to another message.
///
/// # Format
/// ```
///            ↱ @sender — content
/// @sender : text message content
/// ```
pub struct InlineReaction {
  /** ID of the user who send this message */
  sender: u8,
  /** Content of the message */
  content: String,
  /** Time at which the message was send */
  // time: u64, // This should be a DateTime<UTC>

  /** Reaction emojis */
  remojis: Vec<Remoji>,
  /** The message this reaction is targeting */
  target: Box<Msg>,
}

#[derive(Clone, Copy, Serialize, Deserialize, Debug)]
/// # Reaction Emojis
/// All messages can be reacted too using emojis.
///
/// # Format
/// ```
/// @sender : text message content
///            😃 4  😵 6
/// ```
pub struct Remoji {
  charCode: u8,
  quantity: u16,
}
